File: FSUIKit/Classes/FSTapCell.m
////
////  FSTapCell.m
////  ShareEconomy
////
////  Created by FudonFuchina on 16/7/21.
////  Copyright © 2016年 FudonFuchina. All rights reserved.
////
//
//#import "FSTapCell.h"
//
//@implementation FSTapCell {
//    UIView      *_backTapView;
//}
//
//- (instancetype)initWithFrame:(CGRect)frame {
//    self = [super initWithFrame:frame];
//    if (self) {
//        _backTapView = [[UIView alloc] initWithFrame: self.bounds];
//        [self addSubview:_backTapView];
//        
//        UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget: self action: @selector(tapAction)];
//        [_backTapView addGestureRecognizer:tap];
//    }
//    return self;
//}
//
//- (void)tapAction {
//    self->_backTapView.backgroundColor = [UIColor colorWithRed:0 green:0 blue:0 alpha:.5];
//    self->_backTapView.alpha = 0.2;
//    [UIView animateWithDuration:.25 animations:^{
//        self->_backTapView.alpha = 0.1;
//    } completion:^(BOOL finished) {
//        self->_backTapView.alpha = 1;
//        self->_backTapView.backgroundColor = UIColor.clearColor;
//    }];
//    
//    if (_click) {
//        _click(self);
//    }
//}
//
//- (void)layoutSubviews {
//    [super layoutSubviews];
//    _backTapView.frame = self.bounds;
//    _textLabel.frame = CGRectMake(15, 0, self.frame.size.width - 30, self.frame.size.height);
//    _detailTextLabel.frame = CGRectMake(90, 0, self.frame.size.width - 105, self.frame.size.height);
//}
//
//- (UILabel *)textLabel {
//    if (!_textLabel) {
//        _textLabel = [[UILabel alloc] initWithFrame:CGRectMake(15, 0, self.frame.size.width - 30, self.frame.size.height)];
//        _textLabel.textColor = UIColor.blackColor;
//        [self addSubview:_textLabel];
//    }
//    return _textLabel;
//}
//
//- (UILabel *)detailTextLabel {
//    if (!_detailTextLabel) {
//        _detailTextLabel = [[UILabel alloc] initWithFrame:CGRectMake(90, 0, self.frame.size.width - 105, self.frame.size.height)];
//        _detailTextLabel.textAlignment = NSTextAlignmentRight;
//        _detailTextLabel.textColor = UIColor.grayColor;
//        _detailTextLabel.font = [UIFont systemFontOfSize:14];
//        [self addSubview:_detailTextLabel];
//    }
//    return _detailTextLabel;
//}
//
//@end

File: FSUIKit/Classes/Swift/FSTapCell.swift
// FSTapCellS.swift
// Translated from FSTapCell.h/m

import UIKit

public typealias TapCellBlock = (FSTapCell) -> Void

public class FSTapCell: UIView {
    
    private lazy var backTapView: UIView = {
        let v = UIView(frame: bounds)
        addSubview(v)
        
        let tap = UITapGestureRecognizer(target: self, action: #selector(tapAction))
        v.addGestureRecognizer(tap)

        return v
    }()
    
    public lazy var textLabel: UILabel = {
        let lbl = UILabel(frame: CGRect(x: 15, y: 0, width: frame.width - 30, height: frame.height))
        lbl.textColor = .black
        addSubview(lbl)
        return lbl
    }()
    
    public lazy var detailTextLabel: UILabel = {
        let lbl = UILabel(frame: CGRect(x: 90, y: 0, width: frame.width - 105, height: frame.height))
        lbl.textAlignment = .right
        lbl.textColor = .gray
        lbl.font = UIFont.systemFont(ofSize: 14)
        addSubview(lbl)
        return lbl
    }()
    
    public var click: TapCellBlock?
    
    public override init(frame: CGRect) {
        super.init(frame: frame)
        _ = backTapView
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        _ = backTapView
    }
        
    @objc private func tapAction() {
        backTapView.backgroundColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.5)
        backTapView.alpha = 0.2
        UIView.animate(withDuration: 0.25, animations: {
            self.backTapView.alpha = 0.1
        }, completion: { _ in
            self.backTapView.alpha = 1
            self.backTapView.backgroundColor = .clear
        })
        
        click?(self)
    }
    
    public override func layoutSubviews() {
        super.layoutSubviews()
        
        backTapView.frame = bounds
        
        textLabel.frame = CGRect(x: 15, y: 0, width: frame.width - 30, height: frame.height)
        detailTextLabel.frame = CGRect(x: 90, y: 0, width: frame.width - 105, height: frame.height)
    }
}
